File: trip-server/scripts/benchmark/benchmark-sse.ts
/**
 * 场景 2：SSE 流式 chat 压测
 *
 * 用 Node 并发 fetch 测 /api/trip/chat 流式
 * 测：并发数 vs 流完成时间 / chunk 数 / token 数
 */

import { saveResult, percentile, getEnv } from './lib/result-store'

const BASE_URL = process.env.BASE_URL || 'http://localhost:3000'

interface SseMetric {
  concurrency: number
  totalStreams: number
  successStreams: number
  streamDurationsMs: number[]  // 每个流的总耗时
  chunkCounts: number[]
  tokenCounts: { prompt: number; completion: number; total: number }[]
  errors: number
}

async function getToken(): Promise<string> {
  const res = await fetch(`${BASE_URL}/api/user/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ username: 'eval-test', password: 'EvalTest@2026' }),
  })
  const data = (await res.json()) as { data: { token: string } }
  return data.data.token
}

async function runSseStream(token: string, message: string): Promise<{
  durationMs: number; chunks: number; tokens: { prompt: number; completion: number; total: number }; error?: string
}> {
  const start = Date.now()
  const res = await fetch(`${BASE_URL}/api/trip/chat`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
    body: JSON.stringify({ message }),
  })
  if (!res.ok) return { durationMs: Date.now() - start, chunks: 0, tokens: { prompt: 0, completion: 0, total: 0 }, error: `HTTP ${res.status}` }
  const reader = res.body!.getReader()
  const decoder = new TextDecoder()
  let buffer = ''
  let chunks = 0
  let tokens = { prompt: 0, completion: 0, total: 0 }
  while (true) {
    const { value, done } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    // 完整事件以 \n\n 结束
    let idx
    while ((idx = buffer.indexOf('\n\n')) !== -1) {
      const event = buffer.slice(0, idx)
      buffer = buffer.slice(idx + 2)
      const dataLine = event.split('\n').find((l) => l.startsWith('data: '))
      if (!dataLine) continue
      try {
        const data = JSON.parse(dataLine.slice(6))
        if (data.type === 'chunk') chunks++
        if (data.type === 'complete' && data.data?.usage) tokens = data.data.usage
      } catch { /* ignore */ }
    }
  }
  return { durationMs: Date.now() - start, chunks, tokens }
}

async function runConcurrency(concurrency: number, totalStreams: number, token: string): Promise<SseMetric> {
  const messages = [
    '北京 2 天美食', '上海 1 天经典', '成都 3 天慢节奏', '西安 2 天文化',
    '杭州 1 天西湖', '广州 2 天美食', '深圳 1 天现代', '重庆 2 天夜景',
  ]
  const metrics: SseMetric = {
    concurrency, totalStreams, successStreams: 0,
    streamDurationsMs: [], chunkCounts: [], tokenCounts: [], errors: 0,
  }
  let idx = 0
  async function worker() {
    while (idx < totalStreams) {
      const myIdx = idx++
      if (myIdx >= totalStreams) break
      const msg = messages[myIdx % messages.length]
      try {
        const r = await runSseStream(token, msg)
        if (r.error) {
          metrics.errors++
        } else {
          metrics.successStreams++
          metrics.streamDurationsMs.push(r.durationMs)
          metrics.chunkCounts.push(r.chunks)
          metrics.tokenCounts.push(r.tokens)
        }
      } catch (e) {
        metrics.errors++
      }
    }
  }
  await Promise.all(Array.from({ length: concurrency }, () => worker()))
  return metrics
}

async function main() {
  console.log('[sse] 启动 SSE 流式压测...')
  const token = await getToken()

  const results: SseMetric[] = []
  for (const conc of [1, 5, 10, 20]) {
    console.log(`[sse] 并发数: ${conc}`)
    const m = await runConcurrency(conc, 20, token)
    results.push(m)
    console.log(`  完成 ${m.successStreams}/${m.totalStreams} 流，P50=${percentile(m.streamDurationsMs, 50)}ms P99=${percentile(m.streamDurationsMs, 99)}ms`)
  }

  saveResult('sse-results', { scenario: 'sse', env: getEnv(), results })
  console.log('[sse] 完成')
}

main().catch((e) => { console.error(e); process.exit(1) })

File: trip-server-py/scripts/benchmark_sse.py
"""SSE 流式压测脚本（对应 Node 版 benchmark-sse.ts）

测试端点：POST /api/trip/chat
测试 4 个并发级别：1, 5, 10, 20
每个级别发送 20 个 SSE 流请求，统计耗时、chunk 数、token 用量。
"""

import asyncio
import json
import os
import time

import httpx

from benchmark_lib.auth import get_auth_header, BASE_URL
from benchmark_lib.store import get_env, percentile, save_result

TOTAL_STREAMS = int(os.getenv("SSE_TOTAL_STREAMS", "20"))
CONCURRENCY_LEVELS = [int(x) for x in os.getenv("SSE_CONCURRENCY_LEVELS", "1,5,10,20").split(",")]

# 8 条测试消息（与 Node 版一致）
TEST_MESSAGES = [
    "北京 2 天美食",
    "上海 1 天经典",
    "成都 3 天慢节奏",
    "西安 2 天文化",
    "杭州 1 天西湖",
    "广州 2 天美食",
    "深圳 1 天现代",
    "重庆 2 天夜景",
]


async def run_sse_stream(client: httpx.AsyncClient, message: str,
                         headers: dict) -> dict:
    """运行单个 SSE 流请求"""
    start = time.monotonic()
    chunks = 0
    tokens = {}

    try:
        async with client.stream(
            "POST", "/api/trip/chat",
            json={"message": message},
            headers=headers,
            timeout=120,
        ) as resp:
            if resp.status_code != 200:
                return {"error": f"HTTP {resp.status_code}", "durationMs": 0}

            async for line in resp.aiter_lines():
                if line.startswith("data: "):
                    try:
                        data = json.loads(line[6:])
                        ev_type = data.get("type")
                        if ev_type == "chunk":
                            chunks += 1
                        elif ev_type == "complete":
                            usage = data.get("data", {}).get("usage", {})
                            tokens = {
                                "prompt": usage.get("prompt", 0),
                                "completion": usage.get("completion", 0),
                                "total": usage.get("total", 0),
                            }
                    except json.JSONDecodeError:
                        pass

            duration = (time.monotonic() - start) * 1000
            return {"durationMs": duration, "chunks": chunks, "tokens": tokens}

    except Exception as e:
        return {"error": str(e), "durationMs": (time.monotonic() - start) * 1000}


async def run_concurrency(client: httpx.AsyncClient, headers: dict,
                          concurrency: int, total: int) -> dict:
    """以指定并发级别运行 SSE 测试"""
    sem = asyncio.Semaphore(concurrency)
    results = []

    async def _run(msg: str):
        async with sem:
            return await run_sse_stream(client, msg, headers)

    tasks = [_run(TEST_MESSAGES[i % len(TEST_MESSAGES)]) for i in range(total)]
    results = await asyncio.gather(*tasks)

    durations = [r["durationMs"] for r in results if "durationMs" in r and r["durationMs"] > 0]
    chunks = [r["chunks"] for r in results if "chunks" in r]
    token_counts = [r["tokens"] for r in results if "tokens" in r and r["tokens"]]
    errors = sum(1 for r in results if "error" in r and r["error"])

    metric = {
        "concurrency": concurrency,
        "totalStreams": total,
        "successStreams": total - errors,
        "streamDurationsMs": durations,
        "chunkCounts": chunks,
        "tokenCounts": token_counts,
        "errors": errors,
    }

    p50 = percentile(durations, 50) / 1000 if durations else 0
    p99 = percentile(durations, 99) / 1000 if durations else 0
    print(f"  concurrency={concurrency}: success={total - errors}/{total}, "
          f"P50={p50:.1f}s, P99={p99:.1f}s")

    return metric


async def main():
    print(f"SSE Benchmark: {len(CONCURRENCY_LEVELS)} concurrency levels × {TOTAL_STREAMS} streams")
    print(f"Base URL: {BASE_URL}")

    print("Getting auth token...")
    headers = get_auth_header(BASE_URL)

    async with httpx.AsyncClient(base_url=BASE_URL, timeout=120) as client:
        all_results = []
        for conc in CONCURRENCY_LEVELS:
            print(f"Running concurrency={conc}...")
            metric = await run_concurrency(client, headers, conc, TOTAL_STREAMS)
            all_results.append(metric)

    result = {
        "scenario": "sse",
        "env": get_env(),
        "results": all_results,
    }

    save_result("sse-results", result)
    print("SSE benchmark complete.")


if __name__ == "__main__":
    asyncio.run(main())
